src/nvim/terminal.c

```diff
@@ -1192,9 +1192,14 @@ static void refresh_scrollback(Terminal *term, buf_T *buf)
     // pending scrollback row into a string and append it just above the visible
     // section of the buffer
     if (((int)buf->b_ml.ml_line_count - height) >= (int)term->sb_size) {
-      // scrollback full, delete lines at the top
-      ml_delete(1, false);
-      deleted_lines(1, 1);
+      // scrollback full, delete 10% of the lines at the top
+      int lines_to_delete = (int)term->sb_size / 10;
+      int i;
+
+      for (i = 0; i < lines_to_delete; ++i) {
+        ml_delete(1, false);
+        deleted_lines(1, 1);
+      }
     }
```

Comment:
maybe this if-block can go out of the while-loop, and the while-loop can become a for-loop. And then `appended_lines` could also be moved out of the loop.